tracer/src/Datadog.Trace/Agent/MessagePack/SpanMessagePackFormatter.cs

```diff
@@ -50,18 +54,70 @@ private SpanMessagePackFormatter()
             _processIdValueBytes = processId > 0 ? MessagePackSerializer.Serialize(processId) : null;
         }
 
+        // IMessagePackFormatter<ArraySegment<Span>> is deprecated, use IMessagePackFormatter<TraceMessageChunk>
+        public int Serialize(ref byte[] bytes, int offset, ArraySegment<Span> value, IFormatterResolver formatterResolver)
+        {
+            var traceContext = value.Array![value.Offset].Context.TraceContext;
+            var traceChunk = new TraceChunkModel(value, traceContext);
+            return Serialize(ref bytes, offset, traceChunk, formatterResolver);
+        }
+
+        public int Serialize(ref byte[] bytes, int offset, TraceChunkModel value, IFormatterResolver formatterResolver)
+        {
+            int originalOffset = offset;
+            var spans = value.Spans;
+            var spanIds = new HashSet<ulong>();
```

Comment:
You could pre-allocate the capacity to avoid resizing operations:

```suggestion
            var spanIds = new HashSet<ulong>(spans.Count);
```

Reply:
thanks

Reply:
Oops, this was added in .NET Framework 4.7.2.

Reply:
I am using the new ctor when available, but fwiw benchmarks show the same heap allocations (64 bytes) when adding 1 span id all the way up to 17 spans id (that's where I stopped) on both .NET Core and .NET Framework.